components/public-api/typescript/src/metrics.ts

```diff
@@ -0,0 +1,411 @@
+                type = "client_stream";
+                break;
+            case MethodKind.BiDiStreaming:
+                type = "bidi_stream";
+                break;
+        }
+        return {
+            type,
+            service: req.service.typeName,
+            method: req.method.name,
+        };
+    };
+
+    return (next) => async (req) => {
+        async function* incrementStreamMessagesCounter<T>(
+            iterable: AsyncIterable<T>,
+            callback: () => void,
+            handleMetrics: boolean,
+        ): AsyncIterable<T> {
+            let status: Code | undefined;
+            try {
+                for await (const item of iterable) {
+                    callback();
+                    yield item;
+                }
+            } catch (e) {
+                const err = ConnectError.from(e);
+                status = err.code;
+                throw e;
+            } finally {
+                if (handleMetrics && !settled) {
+                    stopTimer({ grpc_code: status ? Code[status] : "OK" });
+                    GRPCMetrics.handled({ ...labels, code: status ? Code[status] : "OK" });
+                }
+            }
+        }
+
+        const labels = getLabels(req);
+        GRPCMetrics.started(labels);
+        const stopTimer = GRPCMetrics.startHandleTimer(labels);
+
+        let settled = false;
+        let status: Code | undefined;
+        try {
+            let request: UnaryRequest | StreamRequest;
+            if (!req.stream) {
+                request = req;
+            } else {
+                request = {
+                    ...req,
+                    message: incrementStreamMessagesCounter(
+                        req.message,
+                        GRPCMetrics.sent.bind(GRPCMetrics, labels),
+                        false,
+                    ),
+                };
+            }
+
+            const res = await next(request);
+
+            let response: UnaryResponse | StreamResponse;
+            if (!res.stream) {
+                response = res;
+                settled = true;
+            } else {
+                response = {
+                    ...res,
+                    message: incrementStreamMessagesCounter(
+                        res.message,
+                        GRPCMetrics.received.bind(GRPCMetrics, labels),
+                        true,
+                    ),
+                };
+            }
+
+            return response;
+        } catch (e) {
+            settled = true;
+            const err = ConnectError.from(e);
+            status = err.code;
+            throw e;
+        } finally {
+            if (settled) {
+                stopTimer({ grpc_code: status ? Code[status] : "OK" });
+                GRPCMetrics.handled({ ...labels, code: status ? Code[status] : "OK" });
+            }
+        }
+    };
+}
+
+export class MetricsReporter {
+    private static readonly REPORT_INTERVAL = 10000;
```

Comment:
we can increase it to 1 min and instead do a final flush with https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon if it turns out to be a problem later